### ISSUE TITLE
1.3: Support new table accessibility attributes

### ISSUE THREAD

**Issue by drmacro** [dita-ot/dita-ot#1779]
Proposal 62

This is most important for HTML output but there may be implications for PDF output as well.

**Comment by robander** [comment]
HTML has a very large amount of code today that automatically sets up these values, so that you do not need to maintain them in your source. Thinking the correct approach is to use the values in the document when specified, otherwise, do the automated accessibility. 

FWIW I still have not encountered a table from any of my users where the automated accessibility was not enough, so I will continue to encourage people to use the automated processing for all but the most extreme table examples. (Those can exist, but again, I have not encountered them except in mock-up samples that exist only to show when these attributes can be useful). Maintaining the values manually in DITA topics will be extremely time consuming and error prone, unless your editor does it for you by implementing exactly the same sort of processing already done in DITA-OT.

**Comment by drmacro** [comment]
Oracle insists that even the accessibility features in 1.3 are not sufficient to meet all accessibility requirements, although I am not convinced. One thing they pointed out was that the lack of a built-in table footer in the DITA table model is a problem. I pointed out that you can get the effect by using @outputclass on <row>. I think there may be some misunderstanding of either what the current OT does or what it can be made to do easily enough. They do have some gnarly tables in their docs (these are the docs for Oracle Database, after all) but it's hard to tell how much is a side effect of legacy practices, how much is failure to configure or extend the OT appropriately, and how much is actual missing features.

I see that the current code is generating cell header references--how does it handle row headers?

**Comment by robander** [comment]
Current code handles row headers based on `@rowheaders="firstcol"` (for complex tables) or based on `@keycol` (for `<simpletable>`). Entries in the designated column are marked up with generated IDs using `<th id="gen-id">`, and entries to the right/left in the same row use `@headers` to reference that cell. Spanning cells are supported, just as with tables that use spanned cells in `<thead>`.

This is clearly a case where you can construct tables that need more - maybe the first two rows are headers (impossible with 1.2 in each model), or maybe the last column is a header (impossible in 1.2 with the complex table). That said - the tables I've most often seen to illustrate this usage are less usable than they would be if flipped so that the two header columns ran across the top. That's why I agree it's possible you will need this markup - maybe you have a table where that is in fact the best way to present information - but I haven't run across one in the millions of topics that we've published.

As for footers - you could do that today with `@outputclass` but I'd agree that using interchangeable markup is nicer. Whether that need is just inertia from legacy practices, I really have no way to say. I believe (not certain?) that the OASIS Exchange Model did away with the `<tfoot>` element present in CALS, which sort of indicates to me that they are not required as often as they once were, but I have absolutely no insight into why they were removed. The very early beta versions of DITA in IBM kept the `<tfoot>` element, but it was removed when OASIS formalized 1.0 using that Exchange Model.

**Comment by drmacro** [comment]
The Oracle argument for tfoot is as a container for a legend that comes before the content rows in the source flow (and thus is easily available to screen readers). But again that's in the generated HTML and we can get that from @outputclass on <row>

I have had some publishing clients that had either real table footers or used them for notes or some sort of legend. Might have been worth pushing for restoring tfoot in 1.3 but that ship has sailed.

If I remember tfoot fell off the list for the exchange model because not enough implementations supported it. But of course that was almost 20 years ago now.

**Comment by robander** [comment]
For what it's worth ... my only nervousness about this tracking issue in general comes from the phrase "most important" -- technically true, it's likely more important in HTML than PDF, but I don't think it is important for most documents. I get nervous at nearly all of the descriptions I've seen of this new feature in 1.3, because so many imply that you *could not* make accessible tables in 1.2 and earlier. In fact, simply using `<thead>` and `@rowheader` in DITA 1.0 has been enough for all the real-world tables I've seen in IBM. 

So yes - there may indeed be complex tables that require using the new markup, and if so, it should absolutely be used. I just don't want people to think they *have* to use it. I've talked to other accessibility experts about this, who agree that telling authors they have to maintain this markup by hand will lead to tables that are marked up incorrectly or incompletely. For a table of any complexity, it's hard to get all your references correct, and even harder to maintain them as the table is edited. In addition, if authors think they have to maintain this by hand with the new attributes, they may not even notice the much easier existing markup. So I'll continue to say the new attributes should only be used when really required.

**Comment by keberlein** [comment]
+ 1 to Robert. I think the new table features are really misleading. Most tables need to be radically simplified, not made more complex. I work repeatedly with clients to discuss what belongs in tables, what can be better handled in definition lists, and how complex tables can be reworked. Most complex tables do not make sense to sighted readers -- because the basic information design is inherently faulty -- if so, they NEVER will be comprehensible to folks reliant on screen readers.

**Comment by robander** [comment]
That said ... I am in no way arguing that we should not support the new markup. It should be supported when used. I just think it shouldn't be used very often. :-)

**Comment by drmacro** [comment]
I'm not disagreeing, just reporting what Oracle's accessibility specialists are concerned about. Whether any of Oracle's tables that require use of these features are good design is a different discussion...

**Comment by robander** [comment]
Agreed - different discussion. I'm not trying to say "your tables are wrong", just "most of your tables don't need all the extra work of maintaining these things by hand..."

**Comment by robander** [comment]
I started to work on this and actually discovered a bug in the grammar files for DITA 1.3, with the new `colspec/@rowheader` attribute. I've notified the OASIS TC and hopefully this should be fixed in an errata: https://lists.oasis-open.org/archives/dita/201709/msg00039.html
